### PR TITLE
[https://nvbugs/5567586][feat] Ampere xqa swa specdec for GPT-OSS Eagle3-one-model

### PR DESCRIPTION
**Pull request by jhaotingc**
## Summary by CodeRabbit

- New Features
  - Added sliding-window-aware attention masking, with per-tile masking and early exits when masking is unnecessary.

- Performance
  - Optimized mask computation and packing paths.
  - Improved warmup accuracy by using the device memory clock rate.
  - Updated memory prefetch to the newer CUDA API.

- Tests
  - Updated warmup utilities to accept device clock rate and reflect it in metrics.
  - Adjusted test harness and call sites to use the new warmup interface and prefetch flow.

### Before this PR
GPT-OSS Eagle3-one-model TP=2, H200/H100
```
[11/25/2025-03:48:07] [TRT-LLM] [I] lm-eval gsm8k exact_match,flexible-extract accuracy: 62.09
[11/25/2025-03:48:07] [TRT-LLM] [I] Hypothesis testing report:
===========================================================
= ACCURACY HYPOTHESIS TESTING
===========================================================
Alpha (Type I:  False Positive): 0.050
Beta  (Type II: False Negative): 0.200
Sigma (Standard deviation): 50.000
#Samples: 1319
Higher is better: True
Theta (Minimum detectable effect): 4.841
Reference accuracy: 90.300
Threshold: 87.097
===========================================================
Evaluated accuracy: 62.092
===========================================================
```

### After this PR
GPT-OSS Eagle3-one-model TP=2, H200/H100
```
[11/25/2025-19:39:07] [TRT-LLM] [I] lm-eval gsm8k results (scores normalized to range 0~100):
|Tasks|Version|     Filter     |n-shot|  Metric   |   | Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|------:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |88.8552|±  |0.8668|
|     |       |strict-match    |     5|exact_match|↑  |58.1501|±  |1.3588|

[11/25/2025-19:39:07] [TRT-LLM] [I] lm-eval gsm8k exact_match,flexible-extract accuracy: 88.86
[11/25/2025-19:39:07] [TRT-LLM] [I] Hypothesis testing report:
===========================================================
= ACCURACY HYPOTHESIS TESTING
===========================================================
Alpha (Type I:  False Positive): 0.050
Beta  (Type II: False Negative): 0.200
Sigma (Standard deviation): 50.000
#Samples: 1319
Higher is better: True
Theta (Minimum detectable effect): 4.841
Reference accuracy: 90.300
Threshold: 87.097
===========================================================
Evaluated accuracy: 88.855
===========================================================

|           Tasks            |Version|   Filter   |n-shot|  Metric   |   | Value |   |Stderr|
|----------------------------|------:|------------|-----:|-----------|---|------:|---|-----:|
|gpqa_diamond_cot_zeroshot_aa|      1|strict-match|     0|exact_match|↑  |68.1818|±  |3.3185|

```

GPT-OSS Eagle3-one-model TP=4, RTX PRO 6000
```
[11/25/2025-19:58:11] [TRT-LLM] [I] lm-eval gsm8k results (scores normalized to range 0~100):
|Tasks|Version|     Filter     |n-shot|  Metric   |   | Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|------:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |90.7506|±  | 0.798|
|     |       |strict-match    |     5|exact_match|↑  |63.3813|±  | 1.327|

|           Tasks            |Version|   Filter   |n-shot|  Metric   |   | Value |   |Stderr|
|----------------------------|------:|------------|-----:|-----------|---|------:|---|-----:|
|gpqa_diamond_cot_zeroshot_aa|      1|strict-match|     0|exact_match|↑  |61.6162|±  |3.4649|
```

Perf
image
image

## Test Coverage

```
pytest -s integration/defs/accuracy/test_llm_api_pytorch.py::TestGPTOSS::test_eagle3_2gpus[cutlass-one_model-overlap_scheduler] # Hopper
pytest -s integration/defs/accuracy/test_llm_api_pytorch.py::TestGPTOSS::test_eagle3_4gpus[cutlass-one_model-overlap_scheduler] # RTX PRO 6000
```

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

### CODE REVIEW

**Review by lowsfer — APPROVED** [review]
Please make sure you locally tested special cases like multi-block mode.

**Comment by jhaotingc** [comment]
> Please make sure you locally tested special cases like multi-block mode.

Thanks @lowsfer, for Ampere XQA, the multi-block mode is not enabled yet in TRTLLM. 
But I did ran the XQA unit test. 
Some note is, there were some edge case that fails before I made my changes, those case are stayed failed (due to limited bandwidth 😅) But the overall accuracy ran by pipeline seemed good.

**Review by mikeiovine — APPROVED** [review]

**Comment by farazkh80** [comment]
hey @pengbowang-nv, @jhaotingc recently went OOO until mid november. If the CI passes, could we merge this PR?

> Can you share some examples of these failing tests? Thanks!


I tested the [accuracy](https://nvbugspro.nvidia.com/bug/5567586?commentNumber=7) on RTX6000 MMLU as well and only a 2% drop with spec-dec compared to baseline. I think spec-dec slight accuracy degradation is an on-going discussion. @mikeiovine may have more context.

**Comment by jhaotingc** [comment]
random error, rebasing and retry

File: tensorrt_llm/_torch/attention_backend/trtllm.py [hunk 1]
```diff
@@ -1190,7 +1190,7 @@ def update_spec_dec_param(
                 dtype=torch.int,
                 device='cuda',
             )
-            if get_sm_version() >= 100:
+            if get_sm_version() >= 100 and get_sm_version() != 120:
```
Comment (farazkh80) [inline tensorrt_llm/_torch/attention_backend/trtllm.py]: could we include 121 as well? 

```suggestion
            if get_sm_version() >= 100 and get_sm_version() not in [120, 121]:
```

File: tensorrt_llm/_torch/attention_backend/trtllm.py [hunk 2]
```diff
@@ -465,7 +465,7 @@ def run(
             self.spec_decoding_generation_lengths,
             self.spec_decoding_position_offsets, self.spec_decoding_packed_mask
         ]
-        if get_sm_version() >= 100:
+        if get_sm_version() >= 100 and get_sm_version() != 120:
```
Comment (farazkh80) [inline tensorrt_llm/_torch/attention_backend/trtllm.py]: similar sm checking for these

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py [hunk 3]
```diff
@@ -3976,17 +3978,23 @@ def test_eagle3(self, moe_backend, mocker):
         mocker.patch.object(GPQADiamond, "MAX_OUTPUT_LEN", MAX_OUTPUT_LEN)
         mocker.patch.object(GPQADiamond, "MAX_INPUT_LEN", MAX_INPUT_LEN)
 
-        # https://nvbugs/5590408: 2-Model overlap scheduling has accuracy issue
-        pytorch_config = dict(disable_overlap_scheduler=True,
-                              cuda_graph_config=CudaGraphConfig())
+        if eagle3_one_model:
+            pytorch_config = dict(disable_overlap_scheduler=False,
```
[the remark this answers is not in the document]
Comment (jhaotingc) [reply tests/integration/defs/accuracy/test_llm_api_pytorch.py]: Not that I know of. I recall there's an extensive acc benchmark done in another nvbug that shows overlap scheduler works with one-model.

**Comment by jhaotingc** [comment]
Really an un-related error.

**Comment by jhaotingc** [comment]
@dongfengy do you mind reviewing the PR? This includes some test changes toward `test_eagle3` test? as you added the previous tests. Thank you!

**Review by yuxianq — APPROVED** [review]

File: tests/integration/defs/accuracy/test_llm_api_pytorch.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jhaotingc) [reply tests/integration/defs/accuracy/test_llm_api_pytorch.py]: updated accuracy test in the PR description, overlap scheduler works with one-model (even two-model now)

**Comment by jhaotingc** [comment]
Random GB200 error.

**Review by dongfengy — APPROVED** [review]

**Comment by jhaotingc** [comment]
this time failed at random in multi-gpu test

**Comment by jhaotingc** [comment]
GPT-OSS 2gpu accuracy test runs too slow (> 1 hr) but accuracy is good.
At AWS Re:Invent now, will revise when I get back.

**Comment by jhaotingc** [comment]
multi-gpu node not online

**Review by chzblych — APPROVED** [review]
